File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.py
from collections import defaultdict


def solution(record):
    users = defaultdict(str)
    events = []
    txt = ('님이 들어왔습니다.', '님이 나갔습니다.')
    
    for r in record:
        event, *userinfo = r.split()
        if event == 'Enter':
            events.append((userinfo[0], 0))
            users[userinfo[0]] = userinfo[1]
        elif event == 'Leave':
            events.append((userinfo[0], 1))
        else:
            users[userinfo[0]] = userinfo[1]
    
    return [f'{users[uid]}{txt[i]}' for uid, i in events]

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

public class Solution {
    
    public static String[] solution(String[] record) {
        Map<String, String> users = new HashMap<>();
        List<String[]> events = new ArrayList<>();
        String[] txt = {"님이 들어왔습니다.", "님이 나갔습니다."};

        for (String r : record) {
            String[] tokens = r.split(" ");
            String event = tokens[0];

            if (event.equals("Enter")) {
                String uid = tokens[1];
                String nickname = tokens[2];
                events.add(new String[]{uid, "0"});
                users.put(uid, nickname);
            } else if (event.equals("Leave")) {
                String uid = tokens[1];
                events.add(new String[]{uid, "1"});
            } else {
                String uid = tokens[1];
                String nickname = tokens[2];
                users.put(uid, nickname);
            }
        }

        String[] answer = new String[events.size()];
        for (int i = 0; i < events.size(); i++) {
            String uid = events.get(i)[0];
            int type = Integer.parseInt(events.get(i)[1]);
            answer[i] = users.get(uid) + txt[type];
        }

        return answer;
    }
}